src/status.cc

```diff
@@ -201,16 +323,74 @@ string StatusPrinter::FormatProgressStatus(const char* progress_status_format,
         out += buf;
         break;
 
-        // Percentage
+        // Percentage of edges completed
       case 'p': {
-        int percent = (100 * finished_edges_) / total_edges_;
+        int percent = 0;
+        if (finished_edges_ != 0 && total_edges_ != 0)
+          percent = (100 * finished_edges_) / total_edges_;
         snprintf(buf, sizeof(buf), "%3i%%", percent);
         out += buf;
         break;
       }
 
-      case 'e': {
-        snprintf(buf, sizeof(buf), "%.3f", time_millis_ / 1e3);
+#define FORMAT_TIME_HMMSS(t)                                                \
+  "%" PRId64 ":%02" PRId64 ":%02" PRId64 "", (t) / 3600, ((t) % 3600) / 60, \
+      (t) % 60
+#define FORMAT_TIME_MMSS(t) "%02" PRId64 ":%02" PRId64 "", (t) / 60, (t) % 60
+
+        // Wall time
+      case 'e':  // elapsed, seconds
+      case 'w':  // elapsed, human-readable
+      case 'E':  // ETA, seconds
+      case 'W':  // ETA, human-readable
+      {
+        // Acquisition.
+        double sec;
+        switch (*s) {
+        case 'e':  // elapsed, seconds
+        case 'w':  // elapsed, human-readable
+          sec = time_millis_ / 1e3;
+          break;
+        case 'E':  // ETA, seconds
+        case 'W':  // ETA, human-readable
+          if (time_predicted_percentage_ == 0.0)
+            sec = -1;  // To be printed as "?".
+          else {
+            // So, we know that we've spent time_millis_ wall clock,
+            // and that is time_predicted_percentage_ percent.
+            // How much time will we need to complete 100%?
+            double total_wall_time = time_millis_ / time_predicted_percentage_;
+            // Naturally, that gives us the time remaining.
+            sec = (total_wall_time - time_millis_) / 1e3;
+          }
+          break;
+        }
+        // Display,
+        if (sec < 0)
+          snprintf(buf, sizeof(buf), "?");
+        else {
+          switch (*s) {
+          case 'e':  // elapsed, seconds
+          case 'E':  // ETA, seconds
+            snprintf(buf, sizeof(buf), "%.3f", sec);
+            break;
+          case 'w':  // elapsed, human-readable
+          case 'W':  // ETA, human-readable
+            if (sec >= 60 * 60)
```

Comment:
Hm, but we should probably print hours for both predicates if *either* of them requires them?
It's a bit asymmetrical otherwise...

Reply:
Behavior changing based on whether `%w` and `%W` are both used in the same format would be weird to me…

It could also waste horizontal space with long descriptions and end up eliding out anyways.

Reply:
> based on whether `%w` and `%W` are both used in the same format

That's not what i said. I wouldn't consider whether or not they are **both** present,
but merely the fact that one of them would be printed with hours,
regardless of whether or not said predicate is actually in the format string.

Reply:
That's fine; the calculations would need to be performed ahead of time instead of as-needed when the format specifier is found then.

Reply:
Yay, now that is something i can live with :)